### DISCUSSION TITLE
Mission Graph

### DISCUSSION THREAD

**Discussion by Arcadianer** [Thorium-Sim/thorium-nova#56] [Ideas]
As discussed in #47 and in the previous virtual meeting, a graph can be utilized to design and run a mission.

Because designing a Graph system to structure narrative content and somehow keep it simple enough for the average person is quite a task, I would like to open a discussion on how this kind of thing should be designed.

Throw some ideas around and let's see what sticks.

Let's start by asking the question:

**What kind/categories of nodes would we need ?**
----

**Triggers**:
Nodes that are evaluated constantly or upon an action and return a boolean if a certain condition is met.

**Inputs**:
Nodes that will execute some action. (Like spawn a ship or tell a NPC to move into some space with warp)

**Operations**:
Nodes that will perform some kind of operation on data. Be it mathematical or just boolean conditions.

These are just some ideas. So what kind of other nodes can you think of.

**Comment by alexanderson1993** [comment]
Just to clarify, I think the plan at this point is to scale back the "mission graph" into much smaller graphs for individual milestones on the mission that are executed independently. See https://github.com/Thorium-Sim/thorium-nova/discussions/47#discussioncomment-1508862.

Even still, this is a relevant discussion since we will still need these nodes.

One small edit to your original post: "actions" should be called "inputs", for consistency with the glossary.

My contributions:

**Entity Filters**:
Nodes which filter the list of ECS entities to find a specific entity.

**Component Getters**:
Nodes which take an entity and component ID as inputs and return the values of that component. 

**Export/Import**:
The export node takes any value and a variable name and makes it available to the import node elsewhere in the graph.

**Generators**:
Nodes which generate some value automatically, like a random number, a random string, a random phrase, etc.

**List Operators**:
Nodes for doing things with lists, like grabbing a specific index, grabbing a random item from the list, etc.

I'm sure there are more that I'll add as I think of them.

**Reply by Froginator45** [reply]
**Alarms:**
When triggered will give the flight director a notification. (may just be built into the various other operations instead of being a stand-alone item.

**Timers:**
Has a trigger, length, and reset inputs. With standard "done" output.

**Counters:**
Counts the number of times triggered. Will output when a predetermined max has been reached. Includes reset input.

**Messages (and other job functions):**
Used to send a message to the communications officer. (Other job functions would also be included: engineering tasks, security tasks, etc.)
I know this was already partially covered by everyone and probably obvious but I feel the need to stress it.

**Loops:**
For and While. Not sure if Do While are needed.

Just a few of what I could think of that might be useful. Others that could be included but probably not as useful are more advanced math functions and PID control blocks. (My real world job is coming out here...)
